_posts/2019-09-30-bert-r/bert_r.Rmd

```diff
@@ -0,0 +1,536 @@
+```
+
+> Note: the max input length is 512 and it is extremely computational even for GPU. 
+
+## Load BERT model into R
+
+We can load the BERT model and automatically pad sequences with ```seq_len``` function. This is why using keras-bert^[[Implementation of the BERT. Official pre-trained models could be loaded for feature extraction and prediction](https://github.com/CyberZHG/keras-bert)] is a better choice, it makes the loading process very easy and comfortable.
+
+```{r}
+model = k_bert$load_trained_model_from_checkpoint(
+  config_path,
+  checkpoint_path,
+  training=T,
+  trainable=T,
+  seq_len=seq_length)
+
+```
+
+## Data structure, reading, preparation
+
+The dataset for this post is taken from [Kaggle Jigsaw Unintended Bias in Toxicity Classification competition](https://www.kaggle.com/c/jigsaw-unintended-bias-in-toxicity-classification).
+
+In order to prepare the dataset, we should write a preprocessing function which will read and tokenize data simultaneously. Then, we feed the outputs of the function as input for BERT model.
+
+```{r}
+# tokenize text
+tokenize_fun = function(dataset) {
+  c(indices, target, segments) %<-% list(list(),list(),list())
+  for ( i in 1:nrow(dataset)) {
+    c(indices_tok, segments_tok) %<-% tokenizer$encode(dataset[[DATA_COLUMN]][i], 
+                                                       max_len=seq_length)
+    indices = indices %>% append(list(as.matrix(indices_tok)))
+    target = target %>% append(dataset[[LABEL_COLUMN]][i])
+    segments = segments %>% append(list(as.matrix(segments_tok)))
+  }
+  return(list(indices,segments, target))
+}
+```
+
+```{r}
+# read data
+dt_data = function(dir, rows_to_read){
+  data = data.table::fread(dir, nrows=rows_to_read)
+  c(x_train, x_segment, y_train) %<-% tokenize_fun(data)
+  return(list(x_train, x_segment, y_train))
+}
+```
+
+## Load dataset
+
+As we have already written the preprocess function, at first, it will read data, then add zeros and encode words into indices. Hence, we will have 3 output files:
+
+- __x_train__ is input matrix for BERT
+- __x_segment__ contains zeros for segment embeddings
+- __y_train__ is the output target which we should predict
+
+```{r}
+c(x_train,x_segment, y_train) %<-% 
+dt_data('~/Downloads/jigsaw-unintended-bias-in-toxicity-classification/train.csv',2000)
+```
+
+## Matrix format for Keras-Bert
+
+Data is in the list format. They need to be extracted and transposed. Then, the train and segment matrices should be placed into the list.
+
+```{r}
+train = do.call(cbind,x_train) %>% t()
+segments = do.call(cbind,x_segment) %>% t()
+targets = do.call(cbind,y_train) %>% t()
+
+concat = c(list(train ),list(segments))
+```
+
+## Calculate decay and warmup steps
+
+Adam optimizer with Warmpup helps to slower learning rate at the beginning of the training process. After certain training steps, the learning rate scheduler will gradually be increased because learning new data without warmup can harm BERT model.
```

Comment:
```suggestion
Using the Adam optimizer with warmup helps to lower the learning rate at the beginning of the training process. After certain training steps, the learning rate will gradually be increased, because learning new data without warmup can negatively affect a BERT model.
```